### PR TITLE
feat: Revert v2alpha SLI inlined Data Source changes

### PR DESCRIPTION
**Pull request by nieomylnieja**
## Motivation

The change introduced to v2alpha with this [PR](https://github.com/OpenSLO/OpenSLO/pull/178) which changed how a Data Source is defined in the SLI schema is proving to be problematic implementation wise.

Here's the current shape of the problematic SLI definition as defined in v1 version:

```yaml
thresholdMetric:
  metricSource:
    type: Redshift
    spec:
      region: eu-central-1
      clusterId: metrics-cluster
      databaseName: metrics-db
      query: SELECT value, timestamp FROM metrics WHERE timestamp BETWEEN :date_from AND :date_to
      accessKeyID: accessKey
      secretAccessKey: secretAccessKey
```

And here's the v2alpha proposal:

```yaml
thresholdMetric:
  redshift:
    accessKeyID: accessKey
    secretAccessKey: secretAccessKey
  spec:
    region: eu-central-1
    clusterId: metrics-cluster
    databaseName: metrics-db
    query: SELECT value, timestamp FROM metrics WHERE timestamp BETWEEN :date_from AND :date_to
```

I agree the PR does move into a good direction with splitting the data source details from the query details. However, I'd argue that the slight ease of writing (saves literally just one line of code) improvement we gain is paid by some HEAVY lifting on the implementation side.

## What's the problem?

Mixing compile time unknown fields with predefined fields.

In the proposal we have two fields that are predefined: `spec` and `dataSourceRef`, the latter being an optional field, which does not change the fact that its type is pre-declared.

The Go representation looks like this:

```go
type SLIMetricSpec struct {
	DataSourceRef               string         `yaml:"dataSourceRef,omitempty"`
	DataSourceSpec              map[string]any `yaml:"spec,omitempty"`
	DataSourceConnectionDetails `yaml:",inline"`
}
```

- `encoding/json` does not have builtin inlining capabilities. It requires writing a custom encoding functions for `SLIMetricSpec`.
- `go-yaml/yaml` supports inlining out of the box (only for maps and structs) will suffer nonetheless in a scenario where we want to delay decoding of the inlined fields with something similar to `json.RawMessage`. In this scenario, when we implement `UnmarshalYAML` for a struct serving the same purpose as `json.RawMessage`, the inlined field will be "greedy" and it will store all the other defined fields like `dataSourceRef` or `spec`.

Furthermore, we can't just think about Go and its ecosystem.
OpenSLO schema should available to adoption in other languages as well. Making it over complicated for implementation makes that accessibility much lower.

Code generation utilities like JSON Schema or Cue will simply not work with such mix of defined and undefined fields.

Finally, I strongly believe OpenSLO should strive to be easy to implement.
We won't attract any users if the schema is not first implemented, just by itself it does not provide any value.
By making it slightly easier to write (I'd argue it might not even be much more readable, even for humans) and much harder to implement we're picking the wrong kind of trade-off.

## What do I propose?

Simply reuse Data Source `spec` in a dedicated field:

```yaml
thresholdMetric:
  dataSource:
    type: redshift
    connectionDetails:
      accessKeyID: accessKey
      secretAccessKey: secretAccessKey
  spec:
    region: eu-central-1
    clusterId: metrics-cluster
    databaseName: metrics-db
    query: SELECT value, timestamp FROM metrics WHERE timestamp BETWEEN :date_from AND :date_to
 ```

It's easy to read, allows reusability code-wise and makes it very straightforward to implement.

### CODE REVIEW

**Comment by fourstepper** [comment]
@programmer04 could be interested in this one. From my POV, LGTM, perhaps save for the fact I would either rename `dataSource` to `metricSource` or `metricSourceRef` to `dataSourceRef` :D

**Comment by nieomylnieja** [comment]
@fourstepper take a look at this comment for my rationale on going with `dataSource` over `metricSource` nomenclature: https://github.com/OpenSLO/OpenSLO/issues/162#issuecomment-1248415555

**Comment by fourstepper** [comment]
@nieomylnieja nice! following that, `metricSourceRef` should be renamed to `dataSourceRef` then?

**Comment by fourstepper** [comment]
Cool, that's my bad for not noticing :)

**Review by fourstepper — APPROVED** [review]
LGTM

**Review by fpiwowarczyk — APPROVED** [review]
To make clear our decision making proposing this solution :).

This change was based on how [Persistent Volumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#types-of-persistent-volumes) are working in K8s environment. Each new data source(volume type) can be introduced as a plugin with its own fields that can be used for anything that maintainer wants. Beside that we keep having fields that are needed for all data sources separately. 

We as OpenSLO community won't be ever able to implement spec for all possible data sources, with unique field names etc.  Because of that we wanted to look for some approach that will be extendable and won't be hard to maintain. Another thing was that it felt easier to work with. 

---

Current solution doesn't block us with planning for plugin based approach, imo it makes spec a bit less clear but can live with that, will unblock us to develop faster `sdk` which I think is our current priority. So I think it does make sense to revert this change. Especially because we don't have capacity to [copy k8s approach](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pluginmanager/pluginwatcher/README.md) and it isn't our priority to enable a lot of data sources.

**Comment by nieomylnieja** [comment]
Exactly, not only k8s pre-defines a lot of the popular PV variants it also has the dynamic registry, these variants of Persistent Volume are in fact each a separate object, it's a bit of a different story to what we're trying to do.

Here's an excerpt from their [JSON Schema](https://kubernetesjsonschema.dev/v1.9.7/_definitions.json#/definitions/io.k8s.api.core.v1.PersistentVolumeSpec):

```json
"io.k8s.api.core.v1.PersistentVolumeSpec": {
      "description": "PersistentVolumeSpec is the specification of a persistent volume.",
      "properties": {
        "accessModes": {
          "description": "AccessModes contains all ways the volume can be mounted. More info: https://kubernetes.io/docs/concepts/storage/persistent-volumes#access-modes",
          "type": "array",
          "items": {
            "type": "string"
          }
        },
        "awsElasticBlockStore": {
          "description": "AWSElasticBlockStore represents an AWS Disk resource that is attached to a kubelet's host machine and then exposed to the pod. More info: https://kubernetes.io/docs/concepts/storage/volumes#awselasticblockstore",
          "$ref": "#/definitions/io.k8s.api.core.v1.AWSElasticBlockStoreVolumeSource"
        },
        "azureDisk": {
          "description": "AzureDisk represents an Azure Data Disk mount on the host and bind mount to the pod.",
          "$ref": "#/definitions/io.k8s.api.core.v1.AzureDiskVolumeSource"
        },
        "azureFile": {
          "description": "AzureFile represents an Azure File Service mount on the host and bind mount to the pod.",
          "$ref": "#/definitions/io.k8s.api.core.v1.AzureFilePersistentVolumeSource"
        },
        "capacity": {
          "description": "A description of the persistent volume's resources and capacity. More info: https://kubernetes.io/docs/concepts/storage/persistent-volumes#capacity",
          "type": "object",
          "additionalProperties": {
            "$ref": "#/definitions/io.k8s.apimachinery.pkg.api.resource.Quantity"
          }
        },
        "cephfs": {
          "description": "CephFS represents a Ceph FS mount on the host that shares a pod's lifetime",
          "$ref": "#/definitions/io.k8s.api.core.v1.CephFSPersistentVolumeSource"
        },
```

We'd essentially have to do the same, and to be fair, maybe we will in the future? But now now for sure.
I do see some merit in having predefined fields for all these well known data source types :shrug:

**Comment by fourstepper** [comment]
If I had to guess, it benefits Kubernetes to have this control over the CSIs, as that also perhaps depends on the Kube version you are running and provides built-in "validation" for the user :D

**Pull request merged**
